File: 846.hand-of-straights.py
# @leet start
from typing import Counter, List

# Idea taken from editorial


class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        # O(n) Time, O(n) Space
        if len(hand) % groupSize != 0:
            return False

        c = Counter(hand)
        while c:
            k, v = next(iter(c.items()))
            # Get the root of any sequence
            while c[k - 1] > 0:
                k -= 1

            for j in range(groupSize):
                curr = k + j
                if c[curr] == 0:
                    return False
                c[curr] -= 1
                if c[curr] == 0:
                    del c[curr]

        return True


# @leet end

File: 846.hand-of-straights.cpp
// @leet start
#include <map>
#include <vector>

using std::vector;

class Solution {
public:
  bool isNStraightHand(vector<int> &hand, int groupSize) {
    // TC O(nlogn) for each making and iterating treemap
    // SC O(n) for treemap
    if (hand.size() % groupSize != 0) {
      return false;
    }

    std::map<int, int> map;
    for (const auto &var : hand) {
      ++map[var];
    }

    while (!map.empty()) {
      if (map.size() < groupSize) {
        return false;
      }
      int firstEle = map.begin()->first;

      for (int i = 0; i < groupSize; ++i) {
        if (map[firstEle + i] == 0) {
          return false;
        }
        --map[firstEle + i];
        if (map[firstEle + i] == 0) {
          map.erase(firstEle + i);
        }
      }
    }
    return true;
  }
};
// @leet end
